src/Keboola/DbExtractor/Application.php

```diff
@@ -4,75 +4,83 @@
 
 namespace Keboola\DbExtractor;
 
-use Keboola\Component\Logger;
+use Keboola\Component\BaseComponent;
 use Keboola\DbExtractor\Configuration\ActionConfigDefinition;
+use Keboola\DbExtractor\Configuration\Config;
 use Keboola\DbExtractor\Configuration\ConfigDefinition;
 use Keboola\DbExtractor\Configuration\ConfigRowDefinition;
-use Keboola\DbExtractor\Exception\UserException;
+use Keboola\Component\UserException;
 use Keboola\DbExtractor\Extractor\Extractor;
+use Psr\Log\LoggerInterface;
 use Symfony\Component\Config\Definition\ConfigurationInterface;
 use Symfony\Component\Config\Definition\Exception\Exception as ConfigException;
-use Symfony\Component\Config\Definition\Processor;
-use ErrorException;
 
-class Application
+class Application extends BaseComponent
 {
-    /** @var string */
-    private $action;
-
-    /** @var array */
-    private $parameters;
-
-    /** @var array */
-    private $state;
-
-    /** @var Logger */
-    private $logger;
-
     /** @var Extractor */
     private $extractor;
 
     /** @var ConfigurationInterface */
     private $configDefinition;
 
-    public function __construct(array $config, Logger $logger, array $state = [])
+    /** @var string */
+    private $configDefinitionClassName;
+
+    public function __construct(array $config, LoggerInterface $logger, array $state = [])
     {
-        static::setEnvironment();
+        $this->setConfigDefinitionClassNameByConfig($config);
 
-        $this->action = $config['action'];
-        $this->parameters = $config['parameters'];
-        $this->state = $state;
-        $this->logger = $logger;
+        parent::__construct($logger);
 
-        $extractorFactory = new ExtractorFactory($this->parameters, $this->state);
-        $this->extractor = $extractorFactory->create($this->logger);
+        $extractorFactory = new ExtractorFactory($this->getConfig()->getParameters(), $state);
+        $this->extractor = $extractorFactory->create($this->getLogger());
+    }
 
-        if ($this->action !== 'run') {
-            $this->configDefinition = new ActionConfigDefinition();
-        } elseif (isset($this->parameters['tables'])) {
-            $this->configDefinition = new ConfigDefinition();
+    private function setConfigDefinitionClassNameByConfig(array $config): void
+    {
+        if ($this->getConfigDefinition()) {
+            $this->configDefinitionClassName = get_class($this->getConfigDefinition());
```

Comment:
I'm not sure how this will work exactly.  If a component needs to override with their own config definition classes they'll need to implement `getConfigDefinition()` in their Application class and it will return the appropriate config definition.  Ok, that makes sense now that I write it down:)

Reply:
Yeah thats wrong one. You're able to set custom config definition via `$application->setConfigDefinition()` but ConfigDefinition is getting in construct, co you could set whatever to it with no effect.

Reply:
It's updated right in constructor.